MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java

```diff
@@ -136,9 +142,33 @@ public void process(Bundle messageData) {
                 // Get the next available output buffer and release frame data
                 // TODO: Consider the case when this call times out and returns null
                 bufferData = getAvailableOutputBufferData();
-                timestamp = bufferData.getLong("TIMESTAMP_US");
+                timestampUs = bufferData.getLong("TIMESTAMP_US");
                 bufferIndex = bufferData.getInt("BUFFER_INDEX");
-                lowResDecoder.releaseOutputBuffer(bufferIndex, timestamp * US_TO_NS);
+
+                // TODO: Fix playback speed issues
+                // Compute the delay in render timestamp between the current frame and the previous
+                // frame.
+                long frameDeltaUs = timestampUs - prevTimestampUs;
+                if (frameDeltaUs <= 0) {
+                    frameDeltaUs = 1_000_000 / 30;
+                }
+                Log.d(TAG, "Frame delta: " + frameDeltaUs);
+                // Set the previous timestamp ("zero out" the timestamps) to the current system
+                // timestamp if it has not been set yet (i.e. equals zero).
+                long renderTimestampNs;
+                long currentTimestampNs = System.nanoTime();
+                if (prevRenderTimestampNs == 0) {
+                    renderTimestampNs = currentTimestampNs + frameDeltaUs * US_TO_NS;
```

Comment:
Can we define:

```
long frameDeltaNs = frameDeltaUs * US_TO_NS;
```

Then use frameDeltaNs instead of conversions in a bunch of places ?